File: MIDTERM-LAB-3/Java/MP20.java
import java.io.*;
import java.util.*;

public class MP20 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter dataset file path: ");
        String path = input.nextLine();

        File file = new File(path);
        if (!file.exists() || !file.isFile()) {
            System.out.println("Invalid file path.");
            return;
        }

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String header = br.readLine(); // ヘッダー行
            String[] keys = header.split(",");
            String line;
            List<Map<String, String>> jsonList = new ArrayList<>();

            while ((line = br.readLine()) != null) {
                String[] values = line.split(",");
                Map<String, String> obj = new LinkedHashMap<>();
                for (int i = 0; i < keys.length && i < values.length; i++) {
                    obj.put(keys[i], values[i]);
                }
                jsonList.add(obj);
            }

            // JSON風に出力
            System.out.println("=== CSV to JSON ===");
            System.out.println(jsonList);

        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        }
    }
}

File: MIDTERM-LAB-3/JavaScript/MP20.js
const fs = require('fs');
const readline = require('readline');
const path = require('path');

function validateFilePath(filePath) {
    if (!fs.existsSync(filePath)) {
        console.log("Error: File does not exist.");
        return false;
    }
    if (path.extname(filePath) !== '.csv') {
        console.log("Error: File is not in CSV format.");
        return false;
    }
    return true;
}

function parseCSV(filePath) {
    const data = fs.readFileSync(filePath, 'utf8');
    const lines = data.split('\n').filter(line => line.trim() !== '');
    const headers = lines[0].split(',');
    const records = lines.slice(1).map(line => {
        const values = line.split(',');
        return Object.fromEntries(headers.map((h, i) => [h.trim(), values[i] ? values[i].trim() : null]));
    });
    return records;
}

async function main() {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    let filePath;
    while (true) {
        filePath = await new Promise(resolve => rl.question("Enter dataset file path: ", resolve));
        if (validateFilePath(filePath)) break;
    }
    const records = parseCSV(filePath);
    console.log("=== CSV to JSON ===");
    console.log(JSON.stringify(records, null, 2));
    rl.close();
}

main();
